File: medium/Hash-Table/ransom-note/ransom_note.cpp
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        /*
         * Check if ransomNote can be built from magazine characters.
         * Approach: Hash map frequency count
         * Time Complexity: O(m + n)
         * Space Complexity: O(1)
         */
        unordered_map<char, int> count;

        // Count characters in magazine
        for (char c : magazine) {
            count[c]++;
        }

        // Verify ransomNote
        for (char c : ransomNote) {
            if (count[c] == 0) {
                return false;
            }
            count[c]--;
        }

        return true;
    }
};

File: medium/Hash-Table/ransom-note/ransom_note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        """
        Determine if ransomNote can be constructed from magazine characters.
        Approach: Hash Map (frequency counting)
        Time Complexity: O(m + n)
        Space Complexity: O(1)
        """
        # Count frequency of each character in magazine
        char_count = {}
        for c in magazine:
            char_count[c] = char_count.get(c, 0) + 1

        # Check if ransomNote can be constructed
        for c in ransomNote:
            if char_count.get(c, 0) == 0:
                return False
            char_count[c] -= 1

        return True
